Enable programmatic forwarding from Go apps (relates to #15 and #50)

laher opened the pull request:
I have taken #50 to the next logical step, such that I can call this code directly from another Go app. As well as re-working the ring buffer, I have extracted the functionality into a package `pkg/forward`.

I appreciate this is a broader scope than the original #15 ticket. The motivation is that app deployment is significantly simplified. This change means that there's no need to install oklog on our app servers, no need to alter service definitions, and (in my case) a lot less puppet to deal with.

Please see this working example of how I can use `pkg/forward`, to forward logs programmatically from within a Go app - in this example I'm using logrus, but I haven't needed to write a logrus-specific solution. The same could be done with the stdlib log package (`log.SetOutput(w)`):

https://github.com/laher/oklogrus/blob/master/examples/oklogrus1ps/oklogrus1ps.go#L24:L32

That highlighted code:

	l := logrus.New()
	f := forward.NewBufferedForwarder(urls, "oklogrus: ", 5)
	r, w := io.Pipe()
	go func() {
		err := f.Forward(r)
		if err != nil {
			log.Fatalf("Error forwarding: %v", err)
		}
	}()
	l.Out = w

 * `oklog/cmd/oklog/forward.go` invokes `pkg/forward` as follows:
 
https://github.com/laher/oklog/blob/feature/pkg-forward/cmd/oklog/forward.go#L128:L143

 * Note that I'd like to add some features in the near future, i.e. there ought to be a 'cancel' feature, and I think it would be useful to somehow report on messages dropped - either via prometheus, or maybe even by injecting a 'messages dropped' message into the buffer...

So, if you're keen on this feature, it feels like there's a lot of little decisions to agree on. Maybe a conversation would help to push it forwards at some stage.

If this feature isn't something you'd like to do for the moment, that's also cool, I can either create a separate repository, or even just merge to `laher/oklog` `master` and offer that for the time being.

Cheers, Happy Seasonal egg-related Holiday.

peterbourgon commented on pkg/forward/ringbuffer.go:
```diff
@@ -0,0 +1,87 @@
+			// Drop record
+			// Note: another alternative would have been to move the 'first' value before putting this record onto the buffer, but it's easier and marginally cheaper to just drop the incoming record.
+			return
+		}
```
Ringbuffers should always cache the most recent data, so yeah, you'll want to drop the oldest record and keep the incoming record.

peterbourgon commented on pkg/forward/ringbuffer.go:
```diff
@@ -0,0 +1,87 @@
+// It either sends the record over the channel, adds it to the buffer, or drops the record if the buffer is full.
+func (b *RingBuffer) Put(record string) {
+	select {
+	case b.ch <- record: // no need to even add to the buffer if something is blocking on the channel
```
I don't like this dual-mode implementation, where the records follow one datapath if there happens to be a waiting reader, and another if there doesn't. The Get method in particular is highly complected as a consequence of this design. Instead, all records should flow through the buffer in all cases.

peterbourgon commented on pkg/forward/ringbuffer.go:
```diff
@@ -0,0 +1,87 @@
+	first int          // the index of the first item in the buffer
+	last  int          // the index of the last item in the buffer
+	ch    chan string  // the channel is used to block on the Get method whenever data is unavailable
+	mutex sync.RWMutex // synchronizes changes to buf, first, last
```
IMO the presence of both a chan and a mutex indicates a problem in the design: the chan is itself synchronized with a mutex, so it's duplication of effort, in some sense.

laher commented:
Thank for the helpful feedback @peterbourgon .

 >  Give me a couple days and I'll see about refactoring it a **bit.**

That would be awesome, thanks. All in good time. 

In the meantime I have addressed some issues you raised, added tests and fixed bugs. _Note, I haven't found a way to avoid the channel+mutex yet, but I do agree it's an uncomfortable mixture_.

Changes:

 * The implementation no longer drops incoming records. They always go onto the buffer. So, the buffer contains the most recent records.
 * This revealed some other bugs in the implementation. Fixed and relevant tests added.
 * The channel now purely exists to allow `Get()` to block whenever the buffer is empty. The channel is now a `chan struct{}`, and the message just signifies a 'there's data now, check the buffer'.
 * The 'unblock' from the Put method (in a select/default) is invoked _after_ putting the record onto the buffer.
 * That channel-receive is no longer synchronised by a lock. (oops!)
 * Good news - the benchmarks are showing 444ns instead of about 1000ns previously.

I'm really interested to see what you can propose. Today I also tried a buffer with 2 channels (1 for incoming + 1 for outgoing). I ran a separate goroutine for managing the state with an infinite `for{ select {} }` receiving into the buffer in and sending back out from the buffer, but I couldn't make it work in a straightforward way (i.e. conditionally blocking on the Get method), so I'm parking it for the moment. :/

Ta.

laher commented:
Zoiks, I wrote an alternative implementation with a buffered channel, it's so simple and seems to work just right. See `RingBufferBCH` in `pkg/forward/ringbuffer_bch.go`. It's faster too. 333ns vs 440ns on the put+get benchmark

Funny, I dismissed this idea weeks ago without trying it.

laher commented:
I'm closing this for now. 

I think it was a worthwhile exercise to try this out, but it seems like it's not really close to acceptance, and meanwhile my personal motivation for this feature diminished when I changed jobs this month (we use Kubernetes at my new job, and our log forwarding is already sorted).

Hopefully the ring buffer itself in #50 is closer to resolution, and I'm sure some variant of #62 will be added at some point in the future. I'll leave the fork and branch in place for reference.

Cheers.
